src/main/java/org/webshar/hrms/service/LeaveAllocationService.java

```diff
@@ -160,26 +160,19 @@ private boolean isStartAndEndDateAreOverlappingWithExistingRecordForGivenEmploye
       throws EntityAlreadyExistsException
   {
     List<LeaveAllocation> leaveAllocationList = leaveAllocationRepository
-        .findAllByEmployeeIdAndLeaveTypeIdAndStartDateLessThanEqualAndEndDateGreaterThanEqual(
-            employeeId, leaveTypeId, startDate, endDate);
+        .findAllByStartDateGreaterThanEqualAndStartDateLessThanEqual(startDate, endDate);
 
     leaveAllocationList.addAll(
         leaveAllocationRepository
-            .findAllByEmployeeIdAndLeaveTypeIdAndStartDateGreaterThanEqualAndEndDateLessThanEqual(
-                employeeId, leaveTypeId, startDate, endDate));
-
-    //Need to optimize this currently just ignore this implementation
-//    leaveAllocationList.addAll(
-//        leaveAllocationRepository
-//            .findAllByEmployeeIdAndLeaveTypeIdAndStartDateGreaterThanEqualAndEndDateGreaterThanEqual(
-//                employeeId, leaveTypeId, startDate, endDate));
-//
-//    leaveAllocationList.addAll(
-//        leaveAllocationRepository
-//            .findAllByEmployeeIdAndLeaveTypeIdAndStartDateLessThanEqualAndEndDateLessThanEqual(
-//                employeeId, leaveTypeId, startDate, endDate));
-
-    if (!leaveAllocationList.isEmpty())
+            .findAllByEndDateGreaterThanEqualAndEndDateLessThanEqual(startDate, endDate));
+    leaveAllocationList.addAll(leaveAllocationRepository.findAllByStartDateLessThanEqualAndEndDateGreaterThanEqual(startDate,endDate));
+
+    Optional<LeaveAllocation>  leaveApplicationResult = leaveAllocationList
+        .stream()
+        .filter(leaveApplication -> (leaveApplication.getEmployee().getId() == employeeId) && (leaveApplication.getLeaveType().getId() == leaveTypeId))
+        .findFirst();
+
+    if (leaveApplicationResult.isPresent())
```

Comment:
Need to improve leave overlapping validation logic.
Explore whether this operation can be performed by a single read/query. 
Also, see if the `exists` clause can be used or not instead of fetching full models.